pyfar/signals/deterministic.py

```diff
@@ -256,6 +393,490 @@ def exponential_sweep_time(n_samples, frequency_range, n_fade_out=90,
+        >>>
+        >>> auto_correlation /= pf.dsp.energy(sweep)
+        >>>
+        >>> # plot auto-correlation
+        >>> with pf.plot.context():
+        >>>     plt.plot(np.arange(-2**7, 2**7), auto_correlation)
+        >>>     plt.gca().set_xlim(-2**7, 2**7)
+        >>>     plt.gca().set_xlabel('time lag in samples')
+        >>>     plt.gca().set_ylabel('auto correlation')
+    """
+
+    return _frequency_domain_sweep(
+        n_samples=n_samples,
+        sweep_type='perfect_linear',
+        frequency_range=[0, sampling_rate / 2],
+        bandpass_order=0,
+        start_margin=0,
+        stop_margin=0,
+        fade_in=None,
+        fade_out=None,
+        sampling_rate=sampling_rate,
+        return_group_delay=return_group_delay)
+
+
+def _frequency_domain_sweep(
+        n_samples, sweep_type, frequency_range, bandpass_order,
+        start_margin, stop_margin, fade_in, fade_out, sampling_rate,
+        return_group_delay):
+    """
+    Frequency domain sweep synthesis with arbitrary magnitude response [#]_.
+
+    Non-unique parameters are documented in ``linear_sweep_freq``,
+    ``exponential_sweep_freq``, ``linear_perfect_sweep``, and
+    ``magnitude_spectrum_weighted_sweep``.
+
+    Parameters
+    ----------
+    sweep_type : Signal, string
+        Specify the magnitude response of the sweep.
+
+        signal
+            The magnitude response as :py:class:`~pyfar.classes.audio.Signal`
+            object. If ``signal.n_samples`` is smaller than `n_samples`, zeros
+            are padded to the end of `signal`. Note that `frequency_range` is
+            not required in this case.
+        ``'linear'``
+            Design a sweep with linearly increasing frequency and a constant
+            magnitude spectrum.
+        ``'exponential'``
+            Design a sweep with exponentially increasing frequency. The
+            magnitude decreases by 3 dB per frequency doubling and has constant
+            energy in filters of relative constant bandwidth (e.g. octaves).
+        ``'perfect_linear'``
+            Perfect linear sweep. Note that the parameters `start_margin`,
+            `stop_margin`, and `frequency_range` are not required in this case.
+
+    Returns
+    -------
+    sweep : Signal
+        The sweep signal. The Signal is in the time domain and has the ``none``
+        FFT normalization (see :py:func:`~pyfar.dsp.fft.normalization`). The
+        sweep parameters are written to `comment`.
+    group_delay_sweep : FrequencyData
+        The analytical group delay of the sweep in seconds as a single sided
+        spectrum. Only returned if `return_group_delay` is ``True``.
+
+    References
+    ----------
+    .. [#] S. Müller, P. Massarani. 'Transfer Function Measurement with Sweeps.
+           Directors Cut Including Previously Unreleased Material and some
+           Corrections.' J. Audio Eng. Soc. 2001, 49 (6), 443-471. (all
+           equations referenced in the code refer to this).
```

Comment:
This citation is only suited for non-perfect sweeps, right?

Reply:
yes and no I'd say. It does not mention perfect sweeps if I remember correctly but the method generally also holds for the perfect linear sweep at least.